### PR TITLE
Fix Python wheel installation issues (multiple subtlety linked items)

### PR DESCRIPTION
**Pull request by bmhowe23**
## Description

This PR contains a variety of fixes that all stemmed from the fact that the instructions provided on https://pypi.org/project/cuda-quantum/0.7.0/ (which originate from `python/README.md`) were no longer working after new versions of transitive dependencies were added to conda-forge, and that meant that the `tensornet` backend could no longer be used. The main point of this PR is to get the instructions in a working state again.

## Details

The full list of changes in this PR are:
* Place conda's library path before system library path in case multiple CUDA versions are installed on the host.
* ~~Pin the cuquantum conda-forge version~~ (this was the original solution for the PR, but there was a desire to remove this conda dependency and simply rely on the Python PIP package dependency that was already pinned).
* Remove cuquantum from the Python installation instructions. (e1e84aee205d2bf079a08c26ab7cfcc0d3de1fd8)
  * Since we're removing this from the install instructions, I had to create `tools/cudaq-qpud/cudaq-qpud.py` as a launch script.
  * Updated docs for the new launch script in 2cd4f29fe7e78c66889030f20a69c1d4faeab5ae.
* Fix a Python library initialization bug that was preventing the pre-existing library detection code from working. (91c9cbf9e370ad5e3cdf376907d7fa37efeb2022)
* Fix a lambda capture bug that was only happening with g++ 11.2.1. (246b1328375139dba81aecaa638d7db60c03fe14 and 7fb1dff4a1c1d680fe57e8a1a5ea7a84bf108a5e). This bug appears to have been due to a compiler bug w/ g++ 11.2.1 that we hadn't previously encountered.
* Fix a bug where the sample_async_remote.py example file was accidentally auto-launching `cudaq-qpud` rather than using the ones provided with the `--server` option (a18113d121576ead860d1473c05805545a87a0d5)
* Add a Publishing test to validate the Python wheel instructions every time we push a new PR into main
  * Link to successful run of new test: https://github.com/NVIDIA/cuda-quantum/actions/runs/8715669430/job/23908299709
    * The failed steps were due to artifacts/images expiring, but they are unrelated to the new test.

### CODE REVIEW

File: python/README.md
Comment (amccaskey) [inline python/README.md:43]: forgive my ignorance because I am not very close to this problem and you have dug into it way more than me - but shouldn't we keep ourselves up to date with the latest from cuquantum?

Comment (bmhowe23) [reply python/README.md:43]: Depending on pending discussions with @bettinaheim and @1tnguyen, I think the `cuquantum` piece will eventually come out of this line. But note that our pyproject.toml file has us version locked to 23.10 as well. I think our problems were stemming from the fact that there was a discrepancy between our pyproject.toml file (which locked us to 23.10) and these instructions (which allowed the version to float).

File: python/README.md
Comment (bmhowe23) [inline python/README.md:44]: Note that regardless of the uncertainty about the above change, this change is important because if the user has multiple versions of CUDA installed on the host, this new change will ensure that the CUDA we just installed above (cuda-11.8.0) will take precedence.

File: python/README.md
Comment (bmhowe23) [reply python/README.md]: With my latest push to this PR, this change is reverted, so I'm going to close the issue. But we should consider updating to the latest 24.x version in 0.8. (The underlying issue was a bug in __init__.py.)

File: .github/workflows/publishing.yml
Comment (bmhowe23) [inline .github/workflows/publishing.yml:831]: Need to address this FIXME

File: python/README.md
Comment (leofang) [inline python/README.md:44]: This is really nerve-wrecking. Conda has explicitly suggested users to **NOT** include conda paths in `LD_LIBRARY_PATH` (see, e.g., [here](https://docs.conda.io/projects/conda-build/en/stable/resources/use-shared-libraries.html#shared-libraries-in-macos-and-linux)). Given that NVIDIA is fully committed to make conda-forge a first-class citizen (ex: we attempt to do Day 1+N release on conda-forge for new CTKs, see e.g. [here](https://github.com/conda-forge/cuda-feedstock/issues/23)), may I nudge the team to put serious consideration to sticking to a safer practice for conda on your roadmap?

Comment (bettinaheim) [reply python/README.md:44]: One thing to point out here is that this only modifies LD_LIBRARY_PATH *within this conda environment*, not generally for the shell or session.

File: runtime/nvqir/CircuitSimulator.h
Comment (bmhowe23) [inline runtime/nvqir/CircuitSimulator.h:155]: 2 points to the person that explains to me -- without falling back on a hypothetical compiler bug explanation -- why the Python changes in this PR require this change. And only in gcc11/x86 builds when paired side-by-side with the installer....on some (but not all) operating systems. It worked fine in gcc11/ARM builds (all operating systems) in the "Side-by-side with installer" test, too, so it really was only gc11+x86+side-by-side-with-installer-on-some-operating-systems.

All other standard Docker tests and wheel tests pass with gcc12, clang16, and llvm bootstrapped builds (both before and after this change). 

In any case, if we need to make this change, I think we have similar code in `runtime/nvqir/cutensornet/simulator_mps_register.cpp`, so we should probably put it in that file as well.

I'm open to other ideas if people have them.

Comment (schweitzpgi) [reply runtime/nvqir/CircuitSimulator.h:155]: I didn't confirm this, but it looks a C++ coding bug.

1. There is the `for_each_pauli` iterator that applies a lambda to each element.
1. Each time, a new instance of the lambda is created with an argument `qubitIdx`.
1. When the type is X or Y, a second lambda is instantiated and pushed on the `basisChange` queue.
1. The second lambda captures a reference to `qubitIdx` from the outer lambda.
1. When the `for_each_pauli` ends, all the outer lambdas are no longer live, so the compiler could reclaim them.
1. `basisChange` is still live, still contains the inner lambdas, and the inner lambdas still hold references to the `qubitIdx` arguments that were in the now defunct and possibly reclaimed outer lambdas.

Passing the value of the element by-value to the inner lambda(s) looks like the correct fix then.

Comment (schweitzpgi) [reply runtime/nvqir/CircuitSimulator.h:155]: Trying again. The inner lambdas are causing a reference capture of the reference qubitIds argument. It seems like this should work fine (and it seems to with most of the compilers Ben has tested). But apparently, it is also known to potentially tickle some bugs. Here is a discussion: https://stackoverflow.com/questions/21443023/capturing-a-reference-by-reference-in-a-c11-lambda

File: scripts/validate_wheel.sh
Comment (bmhowe23) [inline scripts/validate_wheel.sh:174]: @bettinaheim - I think this (i.e., updating validate_wheel.sh per these new lines) is the only way we can keep the `cuquantum` conda package out of the Python installation instructions. This means that pip package users would have to do something similar if they ever wanted to manually launch `cudaq-qpud` like we're doing in this test. Please let me know if you had something else in mind.

Comment (bettinaheim) [reply scripts/validate_wheel.sh:174]: I think we can clean this up a bit by including a launch script instead that automatically does the same thing as the __init__ file for the cudaq python package. We can make that script a python file, since I believe it would only be used by python users (other users don't need that script since there all necessary paths need to be set anyway).

Comment (bmhowe23) [reply scripts/validate_wheel.sh:174]: Just to confirm ... this Python-script approach won't work in the C++-only environment. Is that ok?

Comment (bmhowe23) [reply scripts/validate_wheel.sh]: Updated in 1a59a7251 and 2cd4f29fe.

Comment (bmhowe23) [reply scripts/validate_wheel.sh]: > Updated in [1a59a72](https://github.com/NVIDIA/cuda-quantum/commit/1a59a7251193517ab4627e6efc5315558390114c) and [2cd4f29](https://github.com/NVIDIA/cuda-quantum/commit/2cd4f29fe7e78c66889030f20a69c1d4faeab5ae).

Well, it looks like those didn't work. (Failed publishing run is here: https://github.com/NVIDIA/cuda-quantum/actions/runs/8697234380/job/23852549810). So I'll keep investigating.

File: docs/sphinx/releases.rst
Comment (bmhowe23) [inline docs/sphinx/releases.rst:22]: @bettinaheim - do you want this change included in this PR, or is a new change forthcoming such that it should be backed out?

Comment (bmhowe23) [reply docs/sphinx/releases.rst]: This was backed and and covered in a different PR, so I'm resolving this conversation.

**Comment by schweitzpgi** [comment]
Just checking in here. Is this a draft (per the Conda comments) or are you waiting for a review?

**Comment by bmhowe23** [comment]
> Just checking in here. Is this a draft (per the Conda comments) or are you waiting for a review?

This is no longer just a draft. It is ready for review and is the primary PR required for 0.7.1. It needs code owner approval due to the changes to the .github directory. (@bettinaheim)

**Review by bettinaheim — APPROVED** [review]

**Pull request merged**
